### PR TITLE
Add `removeQueryParameters` option

### PR DESCRIPTION
**Pull request by sarslanhan**
that supports removing unwanted query parameters by providing regular
expressions and strings to be tested against.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ URL to normalize.
 
 ##### normalizeProtocol
 
-Type: `boolean`  
+Type: `boolean`
```
Comment (SamVerschueren) [inline readme.md]: Undo remove spaces. Use the `.editorconfig` plugin for your text editor to prevent this.

File: readme.md [hunk 2]
```diff
@@ -54,7 +54,7 @@ normalizeUrl('//sindresorhus.com:80/', {normalizeProtocol: false});
 
 ##### stripFragment
 
-Type: `boolean`  
+Type: `boolean`
```
Comment (SamVerschueren) [inline readme.md]: same as above

File: readme.md [hunk 3]
```diff
@@ -69,7 +69,7 @@ normalizeUrl('sindresorhus.com/about.html#contact', {stripFragment: false});
 
 ##### stripWWW
 
-Type: `boolean`  
+Type: `boolean`
```
Comment (SamVerschueren) [inline readme.md]: same as above

**Comment by SamVerschueren** [comment]
Not sure about the feature though. It should've been discusses in an issue first.

**Comment by sarslanhan** [comment]
We need this feature for one of our projects, so I thought someone else might also find it useful. For now, i can use my fork as a dependency. 

@sindresorhus let me know if you need anything else about this.

File: index.js [hunk 4]
```diff
@@ -86,8 +86,23 @@ module.exports = function (str, opts) {
 		delete urlObj.search;
 	}
 
+	var queryParameters = queryString.parse(urlObj.search) || {};
```
Comment (sindresorhus) [inline index.js]: No need for `|| {}`. It always returns an object.

File: index.js [hunk 5]
```diff
@@ -86,8 +86,23 @@ module.exports = function (str, opts) {
...
+
+	// remove query unwanted parameters
+	if (opts.ignoredQueryParameters && opts.ignoredQueryParameters.length) {
```
Comment (sindresorhus) [inline index.js]: Just use `Array.isArray(opts.ignoredQueryParameters)`

File: index.js [hunk 6]
```diff
@@ -105,3 +120,15 @@ module.exports = function (str, opts) {
 
 	return str;
 };
+
+function testQueryParameter(name, filters) {
+	filters = filters || [];
+
+	for (var i = 0; i < filters.length; i++) {
```
Comment (sindresorhus) [inline index.js]: `Array#filter` would be better

File: index.js [hunk 7]
```diff
@@ -105,3 +120,15 @@ module.exports = function (str, opts) {
 
 	return str;
 };
+
+function testQueryParameter(name, filters) {
+	filters = filters || [];
```
Comment (sindresorhus) [inline index.js]: You've already ensured it's an array earlier.

**Comment by sindresorhus** [comment]
The option name is kinda misleading, it sounds like you're ignoring query params from being normalized, what you're really doing is removing them.

**Comment by sindresorhus** [comment]
I think this makes sense. `utm_medium` is a plague littering URLs. Should not matter anyways. Wonder if it makes sense to default to removing it?

**Comment by sarslanhan** [comment]
What would be your suggestion for the parameter name?

**Comment by sarslanhan** [comment]
And I think this change is related to the following change you provided in #1 

> Removing unused query variables. A page may only expect certain parameters to appear in the query; unused parameters can be removed. Example:
http://www.example.com/display?id=123&fakefoo=fakebar → http://www.example.com/display?id=123
Note that a parameter without a value is not necessarily an unused parameter.

File: index.js [hunk 8]
```diff
@@ -86,8 +86,23 @@ module.exports = function (str, opts) {
 		delete urlObj.search;
 	}
 
+	var queryParameters = queryString.parse(urlObj.search);
+
+	// remove query unwanted parameters
+	if (Array.isArray(opts.ignoredQueryParameters)) {
+		for (var key in queryParameters) {
+			if (!queryParameters.hasOwnProperty(key)) {
```
Comment (SamVerschueren) [inline index.js]: Why `hasOwnProperty`? In the next version of `query-string`, the returned object won't have a prototype and thus this method will fail. I would go for this.

```js
Object.keys(queryParameters).forEach(function (key) {
    if(queryParameters[key] && testQueryParameter(key, opts.ignoredQueryParameters)) {
        delete queryParameters[key];
    }
});
```

I prefer `Object.keys()` above `for..in`, not sure why though. But apart from that, the same logic applies.

Comment (sindresorhus) [inline index.js]: @SamVerschueren I agree, but I would keep it as is so it's easier to convert to a `for of` loop in the future. Can just remove the `hasOwnProperty` when we upgrade to the new query-string.

**Comment by sindresorhus** [comment]
> What would be your suggestion for the parameter name?

I don't know. Surprise me.

> Not sure about making removal of parameters similar to utm* default though.

@SamVerschueren Thoughts?

**Comment by SamVerschueren** [comment]
> Not sure about making removal of parameters similar to utm* default though.

There should be an option to keep `utm*` stuff as well. But I like the idea of removing them by default.

**Comment by sindresorhus** [comment]
@SamVerschueren I was thinking the default could be `[/^utm.+/i]` and you could just pass in an empty array to override it.

What do you think?

**Comment by kevva** [comment]
> @SamVerschueren I was thinking the default could be [/^utm.+/i] and you could just pass in an empty array to override it.

I know it kinda breaks the expected input type but when disabling features I think `false` (or an empty array) is kinda viable.

**Comment by sindresorhus** [comment]
@kevva I used to do that a lot, but I've become more strict about types and would prefer to keep it to just array. `false` isn't any more readable in code than `[]` anyways.

**Comment by SamVerschueren** [comment]
> I was thinking the default could be [/^utm.+/i] and you could just pass in an empty array to override it.

Seems legit!

**Comment by kevva** [comment]
@sindresorhus, yup, suspected that :).

**Pull request closed without merging**